Add unified checkpoint config

DrownFish19 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleNLP/pull/26 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
Others
### Description
<!-- Describe what this PR does -->
Add unified checkpoint config
Configs to unify hybrid parallel checkpoint. Following options are supports:
- [x]   ignore_save_model_weight: do not save model weight when the master weight exits
- [x]   master_weight_compatible:
        1. if the master weight exits, only load when needed
        2. if master weight does not exit, convert model weight to master weight when needed
- [x]   async_save_to_disk: enable asynchronous saving checkpoints to disk
- [x]   enable_all_options: enable all optimization configurations

DrownFish19 committed:
add unified checkpoint config tests

DrownFish19 committed:
add ignore_save_model_weight

DrownFish19 committed:
add master_weight_compatible config

DrownFish19 committed:
fix master_weight_compatible config

DrownFish19 committed:
fix master_weight_compatible config

DrownFish19 committed:
add async_save_to_disk config

DrownFish19 committed:
add checkpoint_compatible config

DrownFish19 committed:
fix checkpoint_compatible config

DrownFish19 committed:
add enable_all_options config

DesmonDay commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -344,6 +359,22 @@ def load_unified_optimizer(model, optimizer, resume_from_checkpoint, safe_serial
+    if "master_weight_compatible" in args.unified_checkpoint_config:
+        if need_master_weights:
+            if has_master_weights:
+                pass
```
不建议代码里面很多的pass，pass主要还是起到占位的作用，以便后续添加新的代码。这块直接 if not has_master_weights跑下面的分支就行。

DesmonDay commented on llm/llama/tests/test_unified_checkpoint_n1c8_config.py:
```diff
@@ -0,0 +1,1275 @@
+        remove_logs()
+        remove_ckpt(pretrain_arguments["output_dir"])
+        train_args = copy.deepcopy(pretrain_arguments)
+        train_args["tensor_parallel_degree"] = 8
```
后续单测里的这些分布式策略的配置，可以用我那个PR里头的 utils.py中的get_pretrain_arguments，稍微简化一下单测。

DesmonDay commented on paddlenlp/trainer/training_args.py:
```diff
@@ -710,6 +710,23 @@ class TrainingArguments:
+                "Configs to unify hybrid parallel checkpoint.\n"
+                "Following options are supports:\n"
+                "- ignore_data_type_error: ignore the data type and convert to fp16, bf16 or fp32\n"
+                "- ignore_save_model_weight: do not save model weight when the master weight exits\n"
```
exists，后面exit都看看是否拼错了

DrownFish19 replied:
已经修改

ZHUI commented on paddlenlp/trainer/training_args.py:
```diff
@@ -710,6 +710,23 @@ class TrainingArguments:
+            "help": (
+                "Configs to unify hybrid parallel checkpoint.\n"
+                "Following options are supports:\n"
+                "- ignore_data_type_error: ignore the data type and convert to fp16, bf16 or fp32\n"
+                "- ignore_save_model_weight: do not save model weight when the master weight exits\n"
+                "- master_weight_compatible: 1. if the master weight exits, only load when needed\n"
+                "                            2. if master weight does not exit, convert model weight to master weight when needed\n"
+                "- checkpoint_compatible: enable loading old checkpoints, new checkpoint will be saved with aunified checkpoint type\n"
+                "- async_save_to_disk: enable asynchronous saving checkpoints to disk\n"
+                "- enable_all_options: enable all optimization configurations\n"
```
@wawltor  辛苦review一下这块的选项配置命名，直接对用户暴露。需要慎重。

ZHUI commented on paddlenlp/trainer/trainer.py:
```diff
@@ -504,8 +504,11 @@ def _load_from_checkpoint(self, resume_from_checkpoint=None):
         if self.args.unified_checkpoint:
-            if resume_from_checkpoint is not None and self.args.dataset_rank == 0:
+            if "checkpoint_compatible" in self.args.unified_checkpoint_config:
+                pass
```
这里的功能不是 pass，这里的功能应该是 检查是新格式还是老格式。老格式pass，新格式需要load

DrownFish19 replied:
更改为检查是否存在老格式文件。
考虑如下：
1. 仅存在老格式文件，必须是原机原并行规则启动，进而确定老格式文件必须存在；
2. 存在老格式文件和新格式文件（同一文件夹，checkpoint-xx），如果采取checkpoint_compatible策略，应该load老格式文件。
3. 仅存在新格式文件（即不存在老格式文件），当前老格式文件完全不存在，即使设置checkpoint_compatible策略，也要执行load_unified_checkpoint

ZHUI replied:
2 我觉得可以不考虑。直接报错。

ZHUI commented on paddlenlp/trainer/trainer.py:
```diff
@@ -2252,8 +2255,9 @@ def _load_optimizer_and_scheduler(self, checkpoint):
         else:
             if self.args.data_parallel_rank == 0:
-                if self.args.unified_checkpoint:
+                if self.args.unified_checkpoint and "checkpoint_compatible" not in self.args.unified_checkpoint_config:
```
同理

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -84,20 +87,28 @@ def save_unified_checkpoint(args, model, output_dir, safe_serialization=False):
 
+    should_save_model_weight = True
+    if "ignore_save_model_weight" in args.unified_checkpoint_config:
+        if args.fp16_opt_level == "O2":
+            should_save_model_weight = False
```
不要用 fp16_opt_level O2 这个 flag。O2可以没有model_weight。你去看optimizer有没有master_weight 这个属性。

DrownFish19 replied:
修改为optimizer._multi_precision

paddle框架中使用_multi_precision参数标志是否存在master_weight，没有其他条件会修改这个参数。

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -873,3 +904,53 @@ def flatten_list(nested_list):
+                state_dict[k] = state_dict.pop(k).cpu().numpy()
+
+    if is_sync:
+        # 同步
```
中文注释不要

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -127,6 +138,11 @@ def load_unified_checkpoint(model, resume_from_checkpoint: str, safe_serializati
     index_filename = PADDLE_WEIGHTS_INDEX_NAME if not safe_serialization else SAFE_WEIGHTS_INDEX_NAME
+    if "ignore_save_model_weight" in args.unified_checkpoint_config:
+        if args.fp16_opt_level == "O2":
+            index_filename = (
+                PADDLE_MASTER_WEIGHTS_INDEX_NAME if not safe_serialization else SAFE_MASTER_WEIGHTS_INDEX_NAME
+            )
```
sharding情况下，原来 weight 可以 local 加载，但是改完之后，可能需要跨机。

需要 check一下是否需要动态加载。

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -344,6 +359,22 @@ def load_unified_optimizer(model, optimizer, resume_from_checkpoint, safe_serial
     if len(resolved_archive_file) > 1:
         resolved_archive_file = tqdm(resolved_archive_file, desc="Loading optimizer shards")
 
+    need_master_weights = args.fp16_opt_level == "O2"
```
同上，这个判断不够。

DrownFish19 replied:
修改为optimizer._multi_precision。原因同上：paddle框架中使用_multi_precision参数标志是否存在master_weight，没有其他条件会修改这个参数。

DrownFish19 committed:
Merge branch 'develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
fix typos

DrownFish19 committed:
update unified checkpoint config

wawltor commented on paddlenlp/trainer/training_args.py:
```diff
@@ -710,6 +710,23 @@ class TrainingArguments:
+                "- master_weight_compatible: 1. if the master weight exists, only load when needed\n"
+                "                            2. if master weight does not exit, convert model weight to master weight when needed\n"
+                "- checkpoint_compatible: enable loading old checkpoints, new checkpoint will be saved with aunified checkpoint type\n"
+                "- async_save_to_disk: enable asynchronous saving checkpoints to disk\n"
```
async_save 看起来就可以，async_save_model ?

DrownFish19 replied:
async_save好一些。不仅是save model，还会save optimizer。

wawltor commented on paddlenlp/trainer/training_args.py:
```diff
@@ -710,6 +710,23 @@ class TrainingArguments:
+                "Configs to unify hybrid parallel checkpoint.\n"
+                "Following options are supports:\n"
+                "- ignore_data_type_error: ignore the data type and convert to fp16, bf16 or fp32\n"
+                "- ignore_save_model_weight: do not save model weight when the master weight exists\n"
```
ignore 改成 skip ？

DrownFish19 replied:
好的，skip更好理解

ZHUI commented on paddlenlp/trainer/training_args.py:
```diff
@@ -710,6 +710,23 @@ class TrainingArguments:
+                "- ignore_save_model_weight: do not save model weight when the master weight exists\n"
+                "- master_weight_compatible: 1. if the master weight exists, only load when needed\n"
+                "                            2. if master weight does not exit, convert model weight to master weight when needed\n"
+                "- checkpoint_compatible: enable loading old checkpoints, new checkpoint will be saved with aunified checkpoint type\n"
```
和 @wawltor 沟通，这个选项默认打开。可以删除。

DrownFish19 committed:
remove chinese typo

DrownFish19 committed:
Merge branch 'PaddlePaddle:develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
update config name

DrownFish19 committed:
Merge branch 'add_unified_checkpoint_config_simple' of github.com:DrownFish19/PaddleNLP into add_unified_checkpoint_config_simple

DrownFish19 committed:
update skip_save_model_weight and master_weight_compatible

DrownFish19 committed:
update typo

DrownFish19 committed:
update typo

DrownFish19 committed:
add update_model_weight_status

SKIP_SAVE_MODEL_WEIGHT (if config):
    step 1: the index_filename is set as master weights in check_unified_checkpoint
    step 2/3: the master weights are loaded from checkpoint
        load_unified_checkpoint_locally() and load_unified_checkpoint_dynamically()
        will load master weights as model weights

DrownFish19 committed:
update checkpoint_compatible

DrownFish19 committed:
update skip_save_model_weight

DrownFish19 committed:
update skip_save_model_weight

DrownFish19 committed:
update checkpoint_compatible

DrownFish19 committed:
update master_weight_compatible

DrownFish19 committed:
skip_save_model_weight

DrownFish19 committed:
Merge branch 'PaddlePaddle:develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
Merge branch 'PaddlePaddle:develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
update n1c8 test cases

DrownFish19 committed:
Merge branch 'add_unified_checkpoint_config_simple' of github.com:DrownFish19/PaddleNLP into add_unified_checkpoint_config_simple

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -77,8 +79,28 @@
+    "- skip_save_model_weight: do not save model weights when the masters weight exist\n"
+    "- master_weight_compatible: 1. if the master weights exist, only load when needed\n"
+    "                            2. if master weights does not exit, convert model weights to master weights when needed\n"
+    "- checkpoint_compatible: enable loading old checkpoints, new checkpoint will be saved with aunified checkpoint type\n"
```
如 zeyang 建议，这个 `checkpoint_compatible` 参数不要了。默认打开，兼容老格式。

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -210,6 +238,9 @@ def _remove_unused_keys(
             )
+
+        # confirm parameter cast is executed on the same device as model
+        state_dict = nested_copy_place(state_dict, place=paddle.framework._current_expected_place())
```
在这里拷贝的话，显存会多一份占用。

ZHUI replied:
看看能不能在 _load_state_dict_into_model 中支持一下，可以加一个忽略类型type的参数。

DrownFish19 replied:
因为safetensor load之后是在cpu上的FP32，并且CPU和GPU两个设备之间的cast结果存在差异。
1. 在load之前确认在同一设备;
2. 如果不在同一设备，才会进行copy，避免显存或内存占用，`inputs = inputs if inputs.place == place else inputs._copy_to(place, False)`

DrownFish19 committed:
Merge branch 'PaddlePaddle:develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
fix check_origin_checkpoint

DrownFish19 committed:
Merge branch 'add_unified_checkpoint_config_simple' of github.com:DrownFish19/PaddleNLP into add_unified_checkpoint_config_simple

DrownFish19 committed:
fix SKIP_SAVE_MODEL_WEIGHT

fix index_filename_path not defined for MASTER_WEIGHTS_INDEX_NAME

DrownFish19 committed:
fix MASTER_WEIGHT_COMPATIBLE

when has_master_weights=False, index_filename_master_weights=None,
os.path.join can not process None.

DrownFish19 committed:
Merge branch 'PaddlePaddle:develop' into add_unified_checkpoint_config_simple

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -391,6 +421,13 @@ def load_unified_optimizer_locally(model, optimizer, resume_from_checkpoint, saf
 
+    # update has_master_weights and index_filename_master_weights
+    # 1. if the master weight exists, only has_master_weights is set True and loaded when needed
+    # 2. if master weight does not exit, convert model weight to master weight when needed
```
```suggestion
    # 2. if master weight does not exist, convert model weight to master weight when needed
```

DrownFish19 committed:
simplify check_origin_checkpoint

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -1510,3 +1560,129 @@ def flatten_list(nested_list):
+        else optimizer
+    )
+    if hasattr(optimizer, "_multi_precision"):
+        return optimizer._multi_precision
```
可以 unwarp_optmizer 递归找到最底层的 optimizer。

DrownFish19 replied:
4种不同的Optimizer采用了不同的名称，此处暂时通过名称直接获取，后续在paddle框架提个PR，统一一下名称。

ZHUI commented on paddlenlp/trainer/trainer.py:
```diff
@@ -550,7 +568,7 @@ def _load_from_checkpoint(self, resume_from_checkpoint=None):
                     # We load the model state dict on the CPU to avoid an OOM error.
                     state_dict = paddle.load(weights_file, return_numpy=True)
-
+                    # state_dict = nested_cast_tensor(state_dict, paddle.get_default_dtype())
```
不建议这里加 **nested_cast_tensor**，这么cast有问题的。

`self._set_state_dict_in_model(state_dict) ` 或者那里更低层的地方，传入一个 ignore dtype error之类的东西吧。

DrownFish19 committed:
make CHECKPOINT_COMPATIBLE as default

DrownFish19 committed:
update select_model_weight_index func

the os.path.isfile is used to identify local model_weight_index, while distributed_isfile is used to identify global model_weight_index.

DrownFish19 committed:
update test case

remove CHECKPOINT_COMPATIBLE

DesmonDay commented on paddlenlp/trainer/training_args.py:
```diff
@@ -714,6 +714,22 @@ class TrainingArguments:
+            "help": (
+                "Configs to unify hybrid parallel checkpoint.\n"
+                "Following options are supports:\n"
+                "- skip_data_type_error: ignore the data type and convert to fp16, bf16 or fp32\n"
```
这个有用吗？没用的话删掉？

DrownFish19 replied:
所有涉及位置都已经删除

DesmonDay commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -77,8 +79,26 @@
+    "- skip_data_type_error: ignore the data type and convert to fp16, bf16 or fp32\n"
+    "- skip_save_model_weight: do not save model weights when the masters weight exist\n"
+    "- master_weight_compatible: 1. if the master weights exist, only load when needed\n"
+    "                            2. if master weights does not exit, convert model weights to master weights when needed\n"
```
exit -> exist

DesmonDay commented on paddlenlp/trainer/trainer.py:
```diff
@@ -505,14 +505,21 @@ def _load_from_checkpoint(self, resume_from_checkpoint=None):
+                use_unified_checkpoint = True
+                if self.check_origin_checkpoint(resume_from_checkpoint):
+                    use_unified_checkpoint = False
+                    logger.info("Loding origin checkpoint, the next ckeckpoint will be saved as unified checkpoint")
```
Loding -> loading, ckeckpoint -> checkpoint

DesmonDay commented on paddlenlp/trainer/trainer.py:
```diff
@@ -2239,7 +2246,14 @@ def _load_optimizer_and_scheduler(self, checkpoint):
+                use_unified_checkpoint = True
+                if self.check_origin_checkpoint(checkpoint):
+                    use_unified_checkpoint = False
+                    logger.info("Loding checkpoint, the next ckeckpoint will be saved as unified checkpoint")
```
同上

DrownFish19 committed:
update typos

DrownFish19 committed:
simplify unified checkpoint cases

DrownFish19 committed:
fix hybrid_parallel_rng_state_tracker

when loading old ckpt with DP, the checkpoint_rng_state does not have hybrid_parallel_rng_state_tracker.

DrownFish19 committed:
fix typos

DrownFish19 committed:
update comments

DrownFish19 committed:
remove repeated test case

DrownFish19 committed:
update test cases

DrownFish19 committed:
remove unused code

DrownFish19 committed:
update test case

DrownFish19 committed:
fix select_model_weight_index

DrownFish19 committed:
remove unused code

DrownFish19 committed:
Merge branch 'develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
fix distributed_gather for v100

DrownFish19 committed:
Merge branch 'PaddlePaddle:develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
add MAX_CONVERT_CONFIGS for test cases

limit the case num of converting  from N1C8 to N2C4 or N2C4 to N1C8

DrownFish19 committed:
Merge branch 'add_unified_checkpoint_config_simple' of github.com:DrownFish19/PaddleNLP into add_unified_checkpoint_config_simple

DrownFish19 committed:
update dynamic test cases

DrownFish19 committed:
remove rtol for MasterWeightCompatible

DrownFish19 committed:
Merge branch 'PaddlePaddle:develop' into add_unified_checkpoint_config_simple

DrownFish19 committed:
fix unified checkpoint for no fp16 and bp16

DrownFish19 committed:
Merge branch 'add_unified_checkpoint_config_simple' of github.com:DrownFish19/PaddleNLP into add_unified_checkpoint_config_simple

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -1510,3 +1557,127 @@ def flatten_list(nested_list):
+    return has_master_weight, index_filename_master_weights
+
+
+def is_need_master_weight(args, optimizer):
```
```python
def unwrap_optimizer(optimizer):
    while hasattr(optimizer, "_inner_opt") or hasattr(optimizer, "_optim"):
        if hasattr(optimizer, "_inner_opt") : 
            optimizer = optimizer._inner_opt
        if hasattr(optimizer, "_optim"):
            optimizer = optimizer._optim

    return optimizer
```

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -1510,3 +1557,127 @@ def flatten_list(nested_list):
+        optimizer = optimizer._optim
+
+    if hasattr(optimizer, "_multi_precision"):
+        return optimizer._multi_precision and (args.bf16 or args.fp16)
```
这个 `and (args.bf16 or args.fp16)`是必须的吗？

ZHUI commented on paddlenlp/trainer/trainer.py:
```diff
@@ -2915,3 +2931,21 @@ def print_config(self, args=None, key=""):
 
         logger.info("")
+
+    def check_origin_checkpoint(self, resume_from_checkpoint):
```
下个pr换了吧，检测新格式。不要这么麻烦

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -1510,3 +1557,127 @@ def flatten_list(nested_list):
+                index_filename_master_weights = (
+                    PADDLE_WEIGHTS_INDEX_NAME if not safe_serialization else SAFE_WEIGHTS_INDEX_NAME
+                )
+                has_master_weight = True
```
给一个 warining，checkpoint中没有master weight，强行使用weigt转master weight。

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -1342,6 +1388,7 @@ def merge_tensor_parallel_with_shard(state_dict, tp_actions, all_filter_keys):
             tensor = state_dict[key]
             if key in tp_actions:
                 ret = distributed_gather(tensor, dst=j, group=tp_group, offload=False)
+                dist.barrier(group=tp_group)
```
记一下 todo吧，这个应该删除掉。影响性能，

ZHUI commented on paddlenlp/trainer/plugins/unified_checkpoint.py:
```diff
@@ -355,13 +381,13 @@ def load_unified_optimizer(args, model, optimizer, resume_from_checkpoint, safe_
 
 
-def load_unified_optimizer_locally(model, optimizer, resume_from_checkpoint, safe_serialization=False):
+def load_unified_optimizer_locally(args, model, optimizer, resume_from_checkpoint, safe_serialization=False):
```
args 里面需要哪些参数？应为这个args里面太多东西了。具体到函数里面，我们可以将需要的拿出来。
